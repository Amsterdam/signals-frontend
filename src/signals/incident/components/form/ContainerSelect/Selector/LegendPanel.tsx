import React from 'react';
import { createGlobalStyle } from 'styled-components';
import { MapPanelContent } from '@amsterdam/arm-core';

import IconList from 'components/IconList/IconList';

interface Item {
  iconUrl: string;
  label: string;
  id: string;
}

export interface LegendPanelProps {
  variant: 'panel' | 'drawer';
  title: string;
  items?: Item[];
}

// Prevent scrollBar on iOS due to navigation bar
const GlobalStyle = createGlobalStyle`
  body { 
    touch-action: none;
    overflow: hidden; 
  }
`;

const LegendPanel: React.FC<LegendPanelProps> = ({ items = [], title, variant }) =>
  <React.Fragment>
    <GlobalStyle />
    <MapPanelContent variant={variant} title={title}>
      <IconList id="legend" items={items} size={40} />
    </MapPanelContent>
  </React.Fragment>;

export default LegendPanel;
