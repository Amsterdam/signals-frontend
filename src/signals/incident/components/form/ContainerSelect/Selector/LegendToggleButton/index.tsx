import LegendToggleButton from './LegendToggleButton';

export default LegendToggleButton;
