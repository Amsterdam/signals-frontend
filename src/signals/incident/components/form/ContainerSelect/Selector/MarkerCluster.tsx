import React, { useMemo } from 'react';
import type { FunctionComponent } from 'react';
import L from 'leaflet';
import MarkerClusterBase from 'components/MarkerCluster';

export interface Props {
  setLayerInstance: React.Dispatch<React.SetStateAction<L.GeoJSON | undefined>>;
}

const MarkerCluster: FunctionComponent<Props> = ({ setLayerInstance }) => {
  const clusterOptions = useMemo<L.MarkerClusterGroupOptions>(
    () => ({
      showCoverageOnHover: false,
      zoomToBoundsOnClick: true,
      chunkedLoading: true,

      iconCreateFunction: /* istanbul ignore next */ cluster => {
        const childCount = cluster.getChildCount();

        return new L.DivIcon({
          html: `<div data-testid="markerClusterIcon"><span>${childCount}</span></div>`,
          className: 'marker-cluster',
          iconSize: new L.Point(40, 40),
        });
      },
    }),
    []
  );
  return <MarkerClusterBase clusterOptions={clusterOptions} setInstance={setLayerInstance} />;
};

export default MarkerCluster;
