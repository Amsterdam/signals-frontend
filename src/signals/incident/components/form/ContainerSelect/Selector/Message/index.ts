export { ZoomMessage, MapMessage } from './Message';
