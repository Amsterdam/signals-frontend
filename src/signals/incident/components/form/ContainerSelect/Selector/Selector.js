import React, { useCallback, useState, useContext } from 'react';
import styled from 'styled-components';
import Button from 'components/Button';
import { themeColor, themeSpacing } from '@amsterdam/asc-ui';
import ContainerSelectContext from '../context';

const Wrapper = styled.div`
  position: relative;
  border: 1px dotted ${themeColor('tint', 'level3')};
  height: ${themeSpacing(40)};
`;

const ButtonBar = styled.div`
  margin: 0;
  position: absolute;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
  display: flex;
`;

const Selector = () => {
  const { update, close } = useContext(ContainerSelectContext);

  const addContainer = useCallback(
    event => {
      event.preventDefault();
      update('PL734');
    },
    [update]
  );

  const removeContainer = useCallback(
    event => {
      event.preventDefault();
      update(null);
    },
    [update]
  );

  return (
    <Wrapper data-testid="containerSelectSelector">
      <ButtonBar>
        <Button onClick={addContainer}>Container toevoegen</Button>
        <Button onClick={removeContainer}>Container verwijderen</Button>
        <Button onClick={close}>Meld deze container/Sluiten</Button>
      </ButtonBar>
    </Wrapper>
  );
};

export default Selector;
