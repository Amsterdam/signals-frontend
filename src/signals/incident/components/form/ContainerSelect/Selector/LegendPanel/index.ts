// SPDX-License-Identifier: MPL-2.0
// Copyright (C) 2021 Gemeente Amsterdam
import LegendPanel from './LegendPanel'

export default LegendPanel
