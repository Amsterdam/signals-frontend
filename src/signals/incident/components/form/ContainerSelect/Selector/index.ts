// SPDX-License-Identifier: MPL-2.0
// Copyright (C) 2020 - 2021 Gemeente Amsterdam
import Selector from './Selector';

export default Selector;
