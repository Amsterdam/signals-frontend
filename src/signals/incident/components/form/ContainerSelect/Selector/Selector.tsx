import React, { useCallback, useContext, useMemo, useState } from 'react';
import type { FunctionComponent } from 'react';
import ReactDOM from 'react-dom';
import styled from 'styled-components';
import type { MapOptions } from 'leaflet';

import { breakpoint, Paragraph, themeColor, themeSpacing } from '@amsterdam/asc-ui';
import { MapPanel, MapPanelContent, MapPanelDrawer, MapPanelProvider } from '@amsterdam/arm-core';
import { SnapPoint } from '@amsterdam/arm-core/lib/components/MapPanel/constants';
import type { ZoomLevel } from '@amsterdam/arm-core/lib/types';
import { useMatchMedia } from '@amsterdam/asc-ui/lib/utils/hooks';
import { Close } from '@amsterdam/asc-assets';
import type { Variant } from '@amsterdam/arm-core/lib/components/MapPanel/MapPanelContext';

import Button from 'components/Button';
import Map from 'components/Map';
import MAP_OPTIONS from 'shared/services/configuration/map-options';

import ContainerSelectContext from 'signals/incident/components/form/ContainerSelect/context';
import LegendToggleButton from './LegendToggleButton';
import LegendPanel from './LegendPanel';
import ViewerContainer from './ViewerContainer';
import ContainerLayer from './WfsLayer/ContainerLayer';
import WfsLayer from './WfsLayer';
import ContainerList from '../ContainerList';
import ZoomMessage from './ZoomMessage';
import useLayerVisible from './useLayerVisible';

const MAP_PANEL_DRAWER_SNAP_POSITIONS = {
  [SnapPoint.Closed]: '90%',
  [SnapPoint.Halfway]: '50%',
  [SnapPoint.Full]: '0',
};
const MAP_PANEL_SNAP_POSITIONS = {
  [SnapPoint.Closed]: '30px',
  [SnapPoint.Halfway]: '400px',
  [SnapPoint.Full]: '100%',
};

const MAP_CONTAINER_ZOOM_LEVEL: ZoomLevel = {
  max: 12,
};

const MapButton = styled(Button)`
  box-shadow: 0 0 0 2px rgba(0, 0, 0, 0.1);
`;

const Wrapper = styled.div`
  position: absolute;
  top: 0;
  left: 0;
  height: 100%;
  width: 100%;
  box-sizing: border-box; // Override box-sizing: content-box set by Leaflet
`;

const StyledMap = styled(Map)`
  height: 100%;
  width: 100%;

  .marker-cluster {
    color: ${themeColor('tint', 'level1')};
    background-color: ${themeColor('tint', 'level1')};
    box-shadow: 1px 1px 1px #666666;

    div {
      width: 32px;
      height: 32px;
      margin-top: 4px;
      margin-left: 4px;
      background-color: ${themeColor('primary')};
    }

    span {
      line-height: 34px;
    }
  }
`;

const StyledContainerList = styled(ContainerList)`
  margin: ${themeSpacing(2)} 0 ${themeSpacing(4)} 0;
`;

const StyledParagraph = styled(Paragraph)`
  margin-bottom: 0;
  font-size: 16px;
  opacity: 0.6;
`;

const EmptySelectionWrapper = styled.div`
  background-color: ${themeColor('tint', 'level2')};
  height: 100px;
  display: flex;
  align-items: center;
  justify-content: center;
  margin: ${themeSpacing(4)} 0;
`;

const ButtonBarStyle = styled.div<{ layerVisible: boolean }>`
  @media screen and ${breakpoint('max-width', 'tabletM')} {
    margin-top: ${({ layerVisible }) => layerVisible && themeSpacing(11)};
  }
`;

const ButtonBar: FunctionComponent<{ zoomLevel: ZoomLevel }> = ({ children, zoomLevel }) => {
  const layerVisible = useLayerVisible(zoomLevel);

  return (
    (
      <ButtonBarStyle data-testid="zoomMessage" layerVisible={layerVisible}>
        {children}
      </ButtonBarStyle>
    ) || null
  );
};

export interface ButtonBarProps {
  zoomLevel: ZoomLevel;
}

const Selector = () => {
  // to be replaced with MOUNT_NODE
  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  const appHtmlElement = document.getElementById('app')!;
  const { selection, location, meta, update, close } = useContext(ContainerSelectContext);
  const [showDesktopVariant] = useMatchMedia({ minBreakpoint: 'tabletM' });
  const { Panel, panelVariant } = useMemo<{ Panel: FunctionComponent; panelVariant: Variant }>(
    () =>
      showDesktopVariant
        ? { Panel: MapPanel, panelVariant: 'panel' }
        : { Panel: MapPanelDrawer, panelVariant: 'drawer' },
    [showDesktopVariant]
  );

  const mapOptions = useMemo<MapOptions>(
    // eslint-disable-next-line @typescript-eslint/no-unsafe-return
    () => ({
      ...MAP_OPTIONS,
      center: location,
      zoomControl: false,
      minZoom: 10,
      maxZoom: 15,
      zoom: 14,
    }),

    [location]
  );

  const [showLegendPanel, setShowLegendPanel] = useState(false);
  const [showSelectionPanel, setShowSelectionPanel] = useState(true);

  const toggleLegend = () => {
    setShowLegendPanel(() => !showLegendPanel);
  };

  const handleLegendCloseButton = () => {
    setShowLegendPanel(false);
    setShowSelectionPanel(true);
  };

  const removeContainer = useCallback<(itemId: string) => void>(
    itemId => {
      update(selection.filter(({ id }) => id !== itemId));
    },
  [update, selection]
  );

  const mapWrapper = (
    <Wrapper data-testid="containerSelectSelector">
      <StyledMap hasZoomControls={showDesktopVariant} mapOptions={mapOptions}>
        <MapPanelProvider
          mapPanelSnapPositions={MAP_PANEL_SNAP_POSITIONS}
          mapPanelDrawerSnapPositions={MAP_PANEL_DRAWER_SNAP_POSITIONS}
          variant={panelVariant}
          initialPosition={SnapPoint.Halfway}
        >
          <ViewerContainer
            topLeft={
              <ButtonBar zoomLevel={MAP_CONTAINER_ZOOM_LEVEL}>
                <LegendToggleButton onClick={toggleLegend} isRenderingLegendPanel={showLegendPanel} />
              </ButtonBar>
            }
            topRight={
              <ButtonBar zoomLevel={MAP_CONTAINER_ZOOM_LEVEL}>
                <MapButton data-testid="selectorClose" variant="blank" onClick={close} size={44} icon={<Close />} />
              </ButtonBar>
            }
          />

          <Panel data-testid={`panel${showDesktopVariant ? 'Desktop' : 'Mobile'}`}>
            {showSelectionPanel && (
              <MapPanelContent variant={panelVariant} title="Kies de container" data-testid="selectionPanel">
                <Paragraph>U kunt meer dan 1 keuze maken</Paragraph>
                {selection.length ? (
                  <StyledContainerList
                    selection={selection}
                    onRemove={removeContainer}
                    featureTypes={meta.featureTypes}
                  />
                ) : (
                  <EmptySelectionWrapper>
                    <StyledParagraph>Maak een keuze op de kaart</StyledParagraph>
                  </EmptySelectionWrapper>
                )}
                <Button onClick={close} variant="primary">
                  Meld deze container{selection.length > 1 ? 's' : ''}
                </Button>
              </MapPanelContent>
            )}

            {showLegendPanel && (
              <LegendPanel
                onClose={handleLegendCloseButton}
                variant={panelVariant}
                title="Legenda"
                items={meta.featureTypes
                  .filter(({ label }) => label !== 'Onbekend') // Filter the unknown icon from the legend
                  .map(featureType => ({
                    label: featureType.label,
                    iconUrl: `data:image/svg+xml;base64,${btoa(featureType.icon.iconSvg)}`,
                    id: featureType.typeValue,
                  }))}
              />
            )}
          </Panel>
        </MapPanelProvider>

        <ZoomMessage zoomLevel={MAP_CONTAINER_ZOOM_LEVEL}>Zoom in om de objecten te zien</ZoomMessage>

        <WfsLayer zoomLevel={MAP_CONTAINER_ZOOM_LEVEL}>
          <ContainerLayer featureTypes={meta.featureTypes} />
        </WfsLayer>
      </StyledMap>
    </Wrapper>
  );
  return ReactDOM.createPortal(mapWrapper, appHtmlElement);
};

export default Selector;
