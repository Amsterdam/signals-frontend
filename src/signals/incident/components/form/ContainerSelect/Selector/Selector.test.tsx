import React from 'react';
import { render, screen } from '@testing-library/react';
import Selector from './Selector';
import fetchMock from 'jest-fetch-mock';
import containersJson from 'utils/__tests__/fixtures/containers.json';
import { contextValue, withContainerSelectContext } from '../ContainerSelectContext.test';
import userEvent from '@testing-library/user-event';

let showDesktopVariant: boolean;
jest.mock('@amsterdam/asc-ui/lib/utils/hooks', () => ({
  useMatchMedia: () => [showDesktopVariant],
}));

describe('signals/incident/components/form/ContainerSelect/Selector', () => {
  beforeEach(() => {
    fetchMock.resetMocks();
    fetchMock.mockResponseOnce(JSON.stringify(containersJson), { status: 200 });
    showDesktopVariant = false;
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  it('should render the component', async () => {
    render(withContainerSelectContext(<Selector />));

    expect(await screen.findByTestId('containerSelectSelector')).toBeInTheDocument();
  });

  it('should call update when adding container', async () => {
    render(withContainerSelectContext(<Selector />));
    expect(contextValue.update).not.toHaveBeenCalled();

    const addContainerButton = await screen.findByText(/containers toevoegen/i);
    userEvent.click(addContainerButton);
    expect(contextValue.update).toHaveBeenCalledWith(expect.any(Array));
  });

  it('should call update when removing container', async () => {
    render(withContainerSelectContext(<Selector />));
    expect(contextValue.update).not.toHaveBeenCalled();

    const removeContainersButton = await screen.findByText(/containers verwijderen/i);
    userEvent.click(removeContainersButton);
    expect(contextValue.update).toHaveBeenCalledWith([]);
  });

  it('should call close when closing the selector', async () => {
    render(withContainerSelectContext(<Selector />));
    expect(contextValue.close).not.toHaveBeenCalled();

    const button = await screen.findByText('Meld deze container');
    userEvent.click(button);
    expect(contextValue.close).toHaveBeenCalled();
  });

  it('should handle close button on legend panel', async () => {
    showDesktopVariant = true;
    render(withContainerSelectContext(<Selector />));

    userEvent.click(await screen.findByText('Legenda'));
    expect(screen.getByTestId('legend-panel')).toBeInTheDocument();

    userEvent.click(screen.getByTitle('Sluit'));
    expect(screen.queryByTestId('legend-panel')).not.toBeInTheDocument();
    expect(screen.queryByTestId('selection-panel')).toBeInTheDocument();
  });

  it('should render legend panel', async () => {
    showDesktopVariant = true;
    render(withContainerSelectContext(<Selector />));

    userEvent.click(await screen.findByText('Legenda'));

    expect(screen.getByTestId('legend-panel')).toBeInTheDocument();
  });

  it('should render selection panel', async () => {
    showDesktopVariant = true;
    render(withContainerSelectContext(<Selector />));

    expect(await screen.findByTestId('selection-panel')).toBeInTheDocument();
  });

  it('should show desktop version on desktop', async () => {
    showDesktopVariant = true;
    render(withContainerSelectContext(<Selector />));

    expect(await screen.findByTestId('panel-desktop')).toBeInTheDocument();
    expect(screen.queryByTestId('panel-mobile')).not.toBeInTheDocument();
  });

  it('should show mobile version on desktop', async () => {
    render(withContainerSelectContext(<Selector />));

    expect(await screen.findByTestId('panel-mobile')).toBeInTheDocument();
    expect(screen.queryByTestId('panel-desktop')).not.toBeInTheDocument();
  });
});
