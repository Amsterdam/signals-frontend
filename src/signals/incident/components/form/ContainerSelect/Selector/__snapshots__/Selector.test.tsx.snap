// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`signals/incident/components/form/ContainerSelect/Selector should render the component 1`] = `
<div
  id="app"
>
  <div
    class="sc-bSFVuW bUrRBV"
  >
    <div
      class="sc-dkAroR bEmxwH sc-jvfriV iruEqK leaflet-container leaflet-grab leaflet-touch-drag"
      data-max-zoom="15"
      data-min-zoom="10"
      data-testid="map"
      style="position: relative;"
      tabindex="0"
    >
      <div
        class="leaflet-pane leaflet-map-pane"
        style="left: 0px; top: 0px;"
      >
        <div
          class="leaflet-pane leaflet-tile-pane"
        >
          <div
            class="leaflet-layer "
            style="z-index: 1;"
          >
            <div
              class="leaflet-tile-container leaflet-zoom-animated"
              style="z-index: 18; left: 0px; top: 0px;"
            >
              <img
                alt=""
                class="leaflet-tile"
                role="presentation"
                src="https://geodata.nationaalgeoregister.nl/tiles/service/wmts/brtachtergrondkaart/EPSG:28992/14/7243/7987.png"
                style="width: 256px; height: 256px; left: -88px; top: -99px;"
              />
            </div>
          </div>
        </div>
        <div
          class="leaflet-pane leaflet-shadow-pane leaflet-zoom-hide"
        />
        <div
          class="leaflet-pane leaflet-overlay-pane"
        >
          <canvas
            height="0"
            style="left: 0px; top: 0px; width: 0px; height: 0px;"
            width="0"
          />
        </div>
        <div
          class="leaflet-pane leaflet-marker-pane leaflet-zoom-hide"
        />
        <div
          class="leaflet-pane leaflet-tooltip-pane"
        />
        <div
          class="leaflet-pane leaflet-popup-pane"
        />
      </div>
      <div
        class="leaflet-control-container"
      >
        <div
          class="leaflet-top leaflet-left"
        />
        <div
          class="leaflet-top leaflet-right"
        />
        <div
          class="leaflet-bottom leaflet-left"
        />
        <div
          class="leaflet-bottom leaflet-right"
        >
          <div
            class="leaflet-control-attribution leaflet-control"
          >
            <a
              href="https://leafletjs.com"
              title="A JS library for interactive maps"
            >
              Leaflet
            </a>
             | Kaartgegevens Â© Kadaster
          </div>
        </div>
      </div>
    </div>
    <div
      class="sc-iIEYCM ksLkFi sc-gHftXq ivHqzF"
    >
      <div
        class="sc-gsBrbv eCwzcl"
      />
      <div
        class="sc-gsBrbv indjRm"
      />
      <div
        class="sc-gsBrbv dcwqgV"
      />
      <div
        class="sc-gsBrbv yHXJV"
      >
        <div
          data-testid="mapZoom"
        >
          <div
            class="ZoomBar-sc-njk1zm iVhwXQ"
          >
            <button
              class="sc-eCssSg kVAZJT ControlButton-sc-1th5d5p gCvTVo"
              size="44"
              title="Inzoomen"
              type="button"
            >
              <span
                class="sc-bdfBwQ kkEKGF"
                size="20"
              >
                <svg
                  focusable="false"
                  viewBox="0 0 32 32"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M32 14H18V0h-4v14H0v4h14v14h4V18h14z"
                    fill-rule="evenodd"
                  />
                </svg>
              </span>
            </button>
            <button
              class="sc-eCssSg kVAZJT ControlButton-sc-1th5d5p gCvTVo"
              size="44"
              title="Uitzoomen"
              type="button"
            >
              <span
                class="sc-bdfBwQ kkEKGF"
                size="20"
              >
                <svg
                  focusable="false"
                  viewBox="0 0 32 32"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M0 13.714h32v4H0z"
                    fill-rule="evenodd"
                  />
                </svg>
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      class="sc-iIEYCM ksLkFi"
    >
      <div
        class="sc-gsBrbv eCwzcl"
      >
        <div
          class="sc-hkwnrn cyvGYv"
        >
          <div>
            <button
              class="sc-eCssSg giFFTI sc-dRPiIx eaQOAo"
            >
              Container toevoegen
            </button>
            <button
              class="sc-eCssSg giFFTI sc-dRPiIx eaQOAo"
            >
              Container verwijderen
            </button>
            <button
              class="sc-eCssSg giFFTI sc-dRPiIx eaQOAo"
            >
              Meld deze container/Sluiten
            </button>
          </div>
          <p
            as="h6"
            class="sc-crrsfI sc-fFubgz decScN iGjRxA"
          >
            Geselecteerd: 
            [,PL734]
          </p>
        </div>
      </div>
      <div
        class="sc-gsBrbv indjRm"
      />
      <div
        class="sc-gsBrbv dcwqgV"
      />
      <div
        class="sc-gsBrbv yHXJV"
      />
    </div>
  </div>
</div>
`;
